File: logic/valuesFromFrameConversion.py
import logic.frameTypesDefinition as frameType
from logic.sharedStructs import RoverBatteryValues
from logic.sharedStructs import EnvironmentalValues
from logic.sharedStructs import SpeedometerValues
from logic.sharedStructs import Frame

class ValuesFromFrame:

    def roverBatteryValues(frame):
        roverBatteryValues = RoverBatteryValues()
        roverBatteryValues.cellVoltage = frame.data1B[0] / 10.0     #We divide by 10 because the initial value was multiplied by 10, to be sent as an integer
        roverBatteryValues.percentage = frame.data1B[1]
        return roverBatteryValues

    def environmentalValues(frame):
        environmentalValues = EnvironmentalValues()
        environmentalValues.temperature = frame.data2B[0]
        environmentalValues.humidity = frame.data2B[1]
        environmentalValues.pressure = frame.data2B[2]
        environmentalValues.altitude = frame.data2B[3]
        return environmentalValues

    def speedometerValues(frame):
        speedometerValues = SpeedometerValues()
        speedometerValues.rpm = frame.data2B[0]
        speedometerValues.metersPerHour = frame.data2B[1]
        speedometerValues.distance = frame.data2B[2]
        return speedometerValues

File: remote_control_and_touchscreen_module/esp32s3_code/src/valuesFromFrameConversion.cpp
#include "valuesFromFrameConversion.h"

//TODO: From a frame, convert the different fields to "values" structs like "rcValue", environmental values, RPI status values, switches values...
//ex.- EnvMonitorValues getValuesFromEnvMonitorFrame()

RcValues rcValuesFromFrame(Frame frame) {

    RcValues rcValues;
    
    rcValues.y1 = frame.data2B[0];
    rcValues.x1 = frame.data2B[1];
    rcValues.y2 = frame.data2B[2];
    rcValues.x2 = frame.data2B[3];
    rcValues.aux1 = frame.data1B[0];
    rcValues.aux2 = frame.data1B[1];
    rcValues.aux3 = frame.data1B[2];
    rcValues.aux4 = frame.data1B[3];
    
    return rcValues;

}

BatteryValues battValuesFromFrame(Frame frame) {

    BatteryValues battValues;
    
    battValues.cellVoltage = frame.data1B[0] / 10.0;  //We divide by 10.0 because the initial value was multiplied by 10, to be sent as an integer
    battValues.percentage = frame.data1B[1];
    
    return battValues;

}

EnvironmentalValues environmentalValuesFromFrame(Frame frame) {

    EnvironmentalValues envValues;
    
    envValues.temperature = frame.data2B[0];
    envValues.humidity = frame.data2B[1];
    envValues.pressure = frame.data2B[2];
    envValues.altitude = frame.data2B[3];
    
    return envValues;

}

SpeedometerValues speedometerValuesFromFrame(Frame frame) {

    SpeedometerValues speedValues;
    
    speedValues.rpm = frame.data2B[0];
    speedValues.metersPerHour = frame.data2B[1];
    speedValues.distance = frame.data2B[2];
    
    return speedValues;

}

RaspberryPiStatusValues rpiStatusValuesFromFrame(Frame frame) {

    RaspberryPiStatusValues rpiStatusValues;
    
    if (frame.data1B[0] == 1) {
        rpiStatusValues.online = true;
    } else {
        rpiStatusValues.online = false;
    }

    if (frame.data1B[1] == 1) {
        rpiStatusValues.cameraOn = true;
    } else {
        rpiStatusValues.cameraOn = false;
    }

    rpiStatusValues.cpuTemperature = frame.data1B[2];
    rpiStatusValues.cpuWorkload = frame.data1B[3];
    
    return rpiStatusValues;

}